internal/services/producer.go

```diff
@@ -65,3 +68,13 @@ func (p *Producer) Stop() {
 	}
 	log.Printf("Producer closed")
 }
+
+func canaryMessageValueToJson(producerID string, index int, timestamp int64) string {
+	m := CanaryMessage{
+		ProducerID: producerID,
+		MessageID:  index,
+		Timestamp:  timestamp,
+	}
+	json, _ := json.Marshal(m)
+	return string(json)
+}
```

Comment:
I feel like there's 2 operations happening here. WDYT about splitting it up as creation vs. conversion?
```
func (p *Producer) NewCanaryMessage() CanaryMessage {
    return CanaryMessage{
		ProducerID: p.config.ProducerClientID,
		MessageID:  p.index,
		Timestamp:  time.Now().UTC().UnixNano() / int64(time.Millisecond)
	}
}
```
and on CanaryMessage struct, add for conversion:
```
func (cm CanaryMessage) Json() string
```

Reply:
I agree but I don't think that the canary message creation needs to be a method provided by a `Producer` instance and even publically available. It's not called outside. I would just have a `createCanaryMessage` or just leaving the code where it is right now. WDYT? 
For sure having the `Json()` method on the `CanaryMessage` struct makes more sense.

Reply:
I changed my mind it would make sense as a kind of private method on `Producer` struct. :-)